common/src/block_interval.rs

```diff
@@ -0,0 +1,255 @@
+use alloy::primitives::B256;
+use alloy::rpc::types::eth::BlockId;
+use alloy::{hex, providers::Provider, transports::Transport};
+use async_stream::try_stream;
+use thiserror::Error;
+use tokio_stream::Stream;
+use tracing::{error, info};
+
+use crate::parsing;
+
+#[derive(Error, Debug, PartialEq)]
+pub enum BlockIntervalError {
+    #[error("invalid block number '{0}'")]
+    InvalidBlockNumber(String),
+    #[error("invalid follow-from range '{0}', should be `lhs..`")]
+    InvalidFollowFromRange(String),
+    #[error("invalid block interval range '{0}'")]
+    InvalidBlockIntervalRange(String),
+    #[error("could not create bounded stream from unbounded follow-from interval")]
+    IntoBoundedStreamError,
+    #[error("could not create unbounded follow-from stream from fixed bounded interval")]
+    IntoUnboundedStreamError,
+    #[error("could not retrieve latest block number from the provider: {0}")]
+    LatestBlockNumberError(String),
+    #[error("invalid block hash '{0}'")]
+    InvalidBlockHash(String),
+}
+
+/// Range of blocks to be processed and proven.
+#[allow(unused)]
```

Comment:
```suggestion
```
We should add new variants when they're actually implemented.
We can track wishlist either in comments on in a github issue